### ISSUE TITLE
Implement Tag Manager snippet toggle in setup flow if there is an existing tag

### ISSUE THREAD

**Issue by kuasha420** [google/site-kit-wp#4934]
This issue is the Tag Manager counterpart to the #4913. 

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

**Relying on the `feature/existing-tag-simplification` branch:**

* The main instructions in the Tag Manager setup flow should no longer include a note on an existing tag (i.e. the `FormInstructions` component should no longer consider whether `hasExistingTag`).
* The Tag Manager `FormInstructions` should furthermore have their copy adjusted as follows:
    * _Please select your Tag Manager account and container below, the snippet will be inserted automatically on your site._ --> _Please select your Tag Manager account and container below. You can change these later in your settings._
    * _Looks like your site is using paired AMP. Please select your Tag Manager account and relevant containers below, the snippets will be inserted automatically on your site._ --> _Looks like your site is using paired AMP. Please select your Tag Manager account and relevant containers below. You can change these later in your settings._
* In case of an existing tag, a toggle for the module's `useSnippet` setting should show _right below_ the dropdowns, accompanied by a small description text (effectively replacing the text that was removed from `FormInstructions` above):
    * The overall appearance of the toggle and text should match the one in the Analytics setup flow ([see Figma](https://www.figma.com/file/1VseUbVsR8GYPa83fLkaKx/Existing-tag-UX-improvement?node-id=408%3A208) and related #4913).
    * If the selected (primary) container ID (regardless of web or AMP, use `getPrimaryContainerID()` from #5044) is the same that the existing tag is for, the text should say: _A tag {existingTagID} for the selected container already exists on the site.  Make sure you remove it if you want to place the same tag via Site Kit, otherwise they will be duplicated._ (where `{existingTagID}` is the existing tag `GTM-...`, which should match the selected container ID)
        * In this case, the toggle should be off by default which should already be the case based on the existing GTM `useExistingTagEffect()`.
    * Otherwise, i.e. if the selected (primary) container ID (regardless of web or AMP, use `getPrimaryContainerID()` from #5044) is different from what the existing tag is for, the text should say: _An existing tag {existingTagID} was found on the page. If you prefer to collect data using that existing tag, please select the corresponding account and property above._ (where `{existingTagID}` is the existing tag `GTM-...`, which should differ from the selected container ID)
        * In this case, the toggle should be on by default which should already be the case based on the existing GTM `useExistingTagEffect()`.

## Implementation Brief

**Any PR for this must be based on and target the `feature/existing-tag-simplification` branch.**

In `assets/js/modules/tagmanager/components/common/FormInstructions.js`:
- Remove the logic of `hasExistingTag` for existing tag case.
- Add a call to `useExistingTagEffect` hook.
- Replace the following copies:
   * _Please select your Tag Manager account and container below, the snippet will be inserted automatically on your site._ --> _Please select your Tag Manager account and container below. You can change these later in your settings._
    * _Looks like your site is using paired AMP. Please select your Tag Manager account and relevant containers below, the snippets will be inserted automatically on your site._ --> _Looks like your site is using paired AMP. Please select your Tag Manager account and relevant containers below. You can change these later in your settings._

### Create and render `UseSnippetSwitch` component variants

As the text for `UseSnippetSwitch` needs to be changed for the usage in the Setup flow, but remain the same for the current usage in the Settings flow, we are going to create two variants, `SetupUseSnippetSwitch` and `SettingsUseSnippetSwitch` in `assets/js/modules/tagmanager/components/common/` directory for use in each case.

 - Refactor `UseSnippetSwitch` component with the following:
    - Remove the logic to display the toggle text.
    - Add a `description` prop so that renders the text in the appropriate place (where the text used to be).

- Create `SettingsUseSnippetSwitch` component with the following logic:
    - Add the text logic extracted from `UseSnippetSwitch` (previously). Render `UseSnippetSwitch` and pass the text in as the `description` prop.
    - In `assets/js/modules/tagmanager/components/settings/SettingsForm.js` replace the existing use of `UseSnippetSwitch` with `SettingsUseSnippetSwitch`.

 - Create `SetupUseSnippetSwitch` component with the following logic:
    - Get the primary container ID using the `getPrimaryContainerID()` selector - which is being implemented in #5044.
    - Get the container ID using the `getExistingTag()` selector.
    - Render  the `UseSnippetSwitch` component with the following cases:
        - If the primary container ID is the same as the (existing) container ID, the `description` text should say: _A tag {existingTagID} for the selected container already exists on the site.  Make sure you remove it if you want to place the same tag via Site Kit, otherwise they will be duplicated._ (where `{existingTagID}` is the existing tag `GTM-...`, which should match the selected container ID).
        - Otherwise, i.e. if the primary container ID is different from the (existing) container ID, the `description` text should say: _An existing tag {existingTagID} was found on the page. If you prefer to collect data using that existing tag, please select the corresponding account and property above._ (where `{existingTagID}` is the existing tag `GTM-...`, which should differ from the selected container ID).

- Render `SetupUseSnippetSwitch` component in `assets/js/modules/tagmanager/components/setup/SetupForm.js` right before `<ContainerNames />`, here:
https://github.com/google/site-kit-wp/blob/1ce59bef0fce0e3b0d331428bad0eae824224679/assets/js/modules/tagmanager/components/setup/SetupForm.js#L183-L184

### Test Coverage

- No new tests are to be added.

## QA Brief
- Manually add Tag Manager tag snippets to the site's `header.php`.
- Start the Tag Manager Setup flow.
- Verify that the snippet switch appears below the dropdowns as per the AC.
- Ensure the snippet switch toggles **OFF** when selecting an account and a container that does match the existing tag.
- Ensure the snippet switch toggles **ON** when selecting an account and a container that does not match the existing tag.
- Ensure the toggle switch text is appropriate for both cases above as per the AC.
- Ensure the snippet switch can still be toggled manually.
- Verify the snippet text conforms to the AC.
- Ensure the settings flow remains the same as today i.e. the existing tag message should be displayed at the top and the toggle message should be the same as today.

### Setup Flow
<img width="1572" alt="Screenshot 2022-04-18 at 12 28 24 PM" src="https://user-images.githubusercontent.com/24408261/163778804-f4c82389-3638-4c1a-b065-32d13fd682f7.png">

### Settings Flow
<img width="1566" alt="Screenshot 2022-04-18 at 12 27 48 PM" src="https://user-images.githubusercontent.com/24408261/163778819-3ac1e318-adbc-4174-889d-252b04234b5b.png">

## Changelog entry

* Include snippet toggle in Tag Manager setup flow whenever there is an existing tag.

**Comment by felixarntz** [comment]
After review with @marrrmarrr and @aaemnnosttv, it looks like placing two similar Tag Manager tags is not a major problem. We should still discourage it, but I think it's less critical to deactivate the placement here than in Analytics.

So the high-level plan here is the following:
* Do not show the toggle.
* Remove logic to deactivate snippet placement automatically based on an existing tag.
* If there is an existing container, inform about it.
* If the existing container is the same one that the user has selected, warn the user and recommend them to either remove the existing tag or not use the Site Kit Tag Manager module.

**Comment by felixarntz** [comment]
@marrrmarrr and I finalized this today. We decided that for now we will place a toggle in this flow, similarly to Analytics. We discussed the possibility of discouraging use of the module when not placing the tag (since it does nothing else), but decided we're not going to go in that direction at this point, as it shouldn't be a major concern to the user.

Having the module can be potentially useful even if they use the existing tag, e.g. it helps clarify that Tag Manager is being used, or the settings could be accessed at any time for example to re-check about the existing tag, or to adjust the current configuration. cc @aaemnnosttv 

Note that I've opened #5044 as a new prerequisite to parts of this issue.

**Comment by tofumatt** [comment]
I'm confused by:

>   - Refactor `UseSnippetSwitch` component with the following:
>     - Remove the logic to display the toggle text.
>     - Add a `description` prop so that renders the text in the appropriate place (where the text used to be).

Which section of text should be removed, is it this? https://github.com/google/site-kit-wp/blob/7423e21ce4065e5cc585ce012be2f755a691a6ca/assets/js/modules/tagmanager/components/common/UseSnippetSwitch.js#L58-L68 

If so, I don't see that text mentioned in the ACs and I don't see where it goes.

Also: `description` seems an odd prop to use here, we can just use `children` to render the extra text I think. 🤔 

There's a lot of mention of what was "previously" someplace in the IB when mentioning what to add/remove, but I think it's making it a bit confusing. Sometimes with largely/purely copy-related IBs we'll avoid duplicating the copy text from ACs to IB, but in this case it might make it easier to read if you add the text to remove/add in various places right into the IB.

This IB is large enough that I think it'll help readability 😅

**Comment by hussain-t** [comment]
Thanks, @tofumatt.

>Which section of text should be removed, is it this?
...
If so, I don't see that text mentioned in the ACs and I don't see where it goes.

The entire `<p>...</p>` element has to be removed.

The AC mentions:
>The overall appearance of the toggle and text should match the one in the Analytics setup flow ([see Figma](https://www.figma.com/file/1VseUbVsR8GYPa83fLkaKx/Existing-tag-UX-improvement?node-id=408%3A208) and related https://github.com/google/site-kit-wp/issues/4913).

As for the setup flow, the toggle text should be changed as provided in the AC which should be the same as in #4913.

However, the AC didn’t mention anything about the setting flow. That makes sense to have the current toggle text for it. Hence, I followed the same approach as #4913 to separate `SetupUseSnippetSwitch` and `SettingsUseSnippetSwitch`. `SettingsUseSnippetSwitch` component should have the removed/extracted `<p>...</p>` from `UseSnippetSwitch`.

>Also: description seems an odd prop to use here, we can just use children to render the extra text I think. 🤔

I intend to keep it consistent with other `UseSnippetSwitch` components:
https://github.com/google/site-kit-wp/blob/580b51d2e44574b15a49fae02c6ed583be3ef390/assets/js/modules/analytics-4/components/common/UseSnippetSwitch.js#L72-L74

IMO, it makes sense to have it as a `message` or `description` prop rather than rendering it as a `children` :)

Let me know what do you think?

>There's a lot of mention of what was "previously" someplace in the IB when mentioning what to add/remove, but I think it's making it a bit confusing. Sometimes with largely/purely copy-related IBs we'll avoid duplicating the copy text from ACs to IB, but in this case, it might make it easier to read if you add the text to remove/add in various places right into the IB.

This is a good point 👍 I will update the IB accordingly.

**Comment by hussain-t** [comment]
>There's a lot of mention of what was "previously" someplace in the IB when mentioning what to add/remove, but I think it's making it a bit confusing. Sometimes with largely/purely copy-related IBs we'll avoid duplicating the copy text from ACs to IB, but in this case it might make it easier to read if you add the text to remove/add in various places right into the IB.

@tofumatt, I have updated the IB to include the text changes.

**Comment by tofumatt** [comment]
IB ✅ 

Thanks for the updates to this IB @hussain-t 👍🏻

**Comment by hussain-t** [comment]
@felixarntz, @aaemnnosttv, does the snippet toggle and the (new) existing message apply for the setup flow only, or for the settings flow as well? 

As for Analytics, we have it only for the setup flow (#4913). However, the existing tag message is displayed on the top in the Analytics settings flow.

<img width="1570" alt="Screenshot 2022-04-14 at 3 12 42 PM" src="https://user-images.githubusercontent.com/24408261/163359412-3377d353-9c9f-44e4-a516-6870c344c934.png">

As for Tag Manager, we are removing the existing tag message from the `FormInstructions` component, which is common for both the setup and settings flows, and having the new existing message for the setup flow _ONLY_. IMO, it makes sense to have the same for both. I would like to hear your thoughts; what would be the best approach?

**Tag Manager settings form without the changes:**
<img width="1570" alt="Screenshot 2022-04-14 at 3 28 25 PM" src="https://user-images.githubusercontent.com/24408261/163362281-406cc46b-2e96-4d97-9621-b7c09956a8c5.png">

Thanks!

**Comment by aaemnnosttv** [comment]
> does the snippet toggle and the (new) existing message apply for the setup flow only, or for the settings flow as well?

@hussain-t the changes to `FormInstructions` would apply to both setup + settings as it's used in both.

Regarding the text for the use snippet toggle – the IB says:

> As the text for `UseSnippetSwitch` needs to be changed for the usage in the Setup flow, but remain the same for the current usage in the Settings flow, we are going to create two variants,

So (if correct) the IB only has details about what the text should be for the setup case, as settings would remain the same. I think there was discussion at some point to potentially show the snippet switch during setup if there was an existing tag, but maybe not for GTM. @felixarntz ?

**Comment by felixarntz** [comment]
@aaemnnosttv 
> I think there was discussion at some point to potentially show the snippet switch during setup if there was an existing tag, but maybe not for GTM.

Yes, the snippet switch should be shown during setup if there is an existing tag, I think that's specified in the ACs.

However, I see the ACs don't provide information on the settings. In the settings, the snippet switch should still be shown at all times, like today.

@hussain-t Looking at your screenshots, please double-check the precise UI. The new toggle and the text below should look similar like the new Analytics toggle per the ACs, but in your screenshot they look very similar to today's toggle with a regular paragraph below. Please follow the Figma design linked in the ACs, or more importantly, use the same approach as the new Analytics setup toggle from #4913 does.

**Comment by hussain-t** [comment]
@felixarntz, @aaemnnosttv, thanks for the clarification on having the toggle snippet and the new message in the settings 👍

 @felixarntz, I assume you're referring to the first (TM) screenshot I posted which has the regular message. To clarify, I mentioned the following text above the screenshot:
>Tag Manager settings form with the changes:

That was my original question "should the new toggle message reflect / the same as the setup flow?" :) 

**Here is the setup flow with the new toggle message that is similar to the Analytics setup:**

<img width="1570" alt="Screenshot 2022-04-15 at 2 59 41 PM" src="https://user-images.githubusercontent.com/24408261/163555714-8977db0b-e9e7-42ca-acfc-a9c9cd5d912e.png">

-------------

**Here is the settings with the new toggle message:**

<img width="1570" alt="Screenshot 2022-04-15 at 3 14 21 PM" src="https://user-images.githubusercontent.com/24408261/163555976-5dc28640-21e2-4792-ad59-c7318885736e.png">

**Comment by felixarntz** [comment]
@hussain-t Actually, taking another look here, we should leave the settings untouched, i.e. keep the toggle the same way as it was before. That also means, can we bring back the text about the existing tag at the top for now, when in settings? This is the same approach that we've taken for Analytics. We can revisit enhancing the settings UI separately. In other words, I think we should probably bring back the original existing tag text in `FormInstructions` with a prop for whether the existing tag information should show there or not.

Basically, the settings should look like they did before, only the setup UI should change based on the ACs. Sorry for the confusion before, but we should really only update the setup UI for now.

**Comment by hussain-t** [comment]
Thanks for the clarification, @felixarntz. That SGTM 👍

**Comment by wpdarren** [comment]
### QA Update: ⚠️ 

@hussain-t  One observation - the AC mentions that it should be styled like the is Analytics toggle/text [Figma design](https://www.figma.com/file/1VseUbVsR8GYPa83fLkaKx/Existing-tag-UX-improvement?node-id=408%3A208) but as you can see from my screenshot its not looking the same.  The notice text is not aligned underneath the toggle text.

This is what I see on Figma (but for Analytics)
![image](https://user-images.githubusercontent.com/73545194/164373223-946720c4-2dcc-4e6e-a01b-0b8405cd3f7e.png)

This is what I see on my test site
![image](https://user-images.githubusercontent.com/73545194/164373601-50d22cc8-87d5-45e7-8577-a65abfec41f1.png)

Could you confirm how it should look?

**Verified:**

**With no existing tag**

- When setting up Tag Manager the text `Please select your Tag Manager account and container below. You can change these later in your settings.` appears. 
- With no existing tag, when the site has AMP installed, and setting up Tag Manager the text `Looks like your site is using paired AMP. Please select your Tag Manager account and relevant containers below. You can change these later in your settings.` appears.

![image](https://user-images.githubusercontent.com/73545194/164370167-a5f3a71a-12cf-44c7-9269-76ef65e8a65b.png)

![image](https://user-images.githubusercontent.com/73545194/164369491-6156ce1f-850e-471b-a3d3-abd80bbbe1c1.png)

**With an existing tag**

- When the selected account/container is the same that the existing tag is for, the text says: `A tag {existingTagID} for the selected container already exists on the site.  Make sure you remove it if you want to place the same tag via Site Kit, otherwise they will be duplicated.`. The toggle is disabled when the container selection is made. 

![image](https://user-images.githubusercontent.com/73545194/164371307-ce5caab8-7f4d-4d31-b265-521193c7f718.png)

- When the selected account/container is different than the existing tag, the text says: `An existing tag {existingTagID} was found on the page. If you prefer to collect data using that existing tag, please select the corresponding account and property above.` The toggle is enabled when the container selection is made. 

![image](https://user-images.githubusercontent.com/73545194/164372938-c1f0021f-eb53-4263-a5ea-64a098b60151.png)
